DOCSP-13871: FAQ page (#226)
mongodb/docs-golang: source/faq.txt

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

> .. tip::

Author: moved first sentence to normal text, left link to I&H in tip

   This page contains frequently asked questions and their corresponding
   answers. If you can't find an answer to your problem on this page,
   see the :ref:`golang-issues-and-help` page for next steps and more
   resources.
   

> How Can I Fix the "WriteXXX() can only write while positioned on an Element or Value but is positioned on a TopLevel" Error?

Author: This is the error message produced by the code example:
`WriteNull can only write while positioned on a Element or Value but is positioned on a TopLevel`. 
So, WriteXXX is a placeholder as the error message depends on the type passed to `bson.Marshal()`.

> How Can I Fix the "WriteXXX() can only write while positioned on an Element or Value but is positioned on a TopLevel" Error?

Reviewer: Issue:
I think we should avoid "XXX" as a placeholder since it's not a widely used that way and doesn't make it more searchable per my comment.

Suggestion:
Choose one of the "Write...()" operations such as "WriteNull" and use that as the title. Then, in the answer to this question, mention some or all of the "Write...()" operations to which this could apply.

> How Can I Fix the "WriteXXX() can only write while positioned on an Element or Value but is positioned on a TopLevel" Error?

Author: changed answer to address specific error but indicated the causes of similar errors (only differ by type)

--------------------------------------------------------------------------------------------------------------------------------

The ``bson.Marshal()`` method requires a parameter that can be decoded
into a BSON Document, such as a ``bson.D`` type. This error occurs
when the input to ``bson.Marshal()`` is something *other* than a BSON
Document.

The following list includes some situations that surface this error:

- You pass a string to ``bson.Marshal()``, causing a ``WriteString`` error.
- You pass a ``null`` to ``bson.Marshal()``, causing a ``WriteNull`` error.

> - You pass a string to ``bson.Marshal()``, causing a ``WriteInt32`` error.

Author: oops, that is a mistake. changed string to integer in the third item.

Many CRUD operations use the ``bson.Marshal()`` method internally,
so you may encounter this error even if you are not directly attempting

> to encode data.

Author: I'm not sure I completely understand what you mean by "seems to imply that their usage was directly attempting to encode data". I think you overall meant that it will be easier for users to understand the two main situations when this error arises: when they directly call `bson.Marshal()` and when they perform a CRUD operation that internally uses this method. Changed wording to reflect this.

The following code produces this error because of how the ``sort``

> variable is initialized:

Author: I reworded to explain what exactly surfaces the error

   opts := options.FindOneAndUpdate().SetSort(sort)

   update := bson.D{{"$inc", bson.D{{"counter", 1}}}}
   result := coll.FindOneAndUpdate(ctx, bson.D{}, update)
   if err := result.Err(); err != nil {
   	log.Fatalf("error getting single result: %v", err)
   }

The following code shows how to correctly initialize the ``sort``
variable as a ``bson.D`` type:

.. code-block:: go

   sort := bson.D{}

How Do I Convert a BSON Document to JSON?
-----------------------------------------

The driver provides a variety of marshaller methods that can be used to
encode a BSON document as JSON, such as the ``MarshalExtJSON()`` method:

.. io-code-block::
   :copyable: true

   .. input::
      :language: go
      :emphasize-lines: 3

      doc := bson.D{{"hello", "world"}}

      jsonBytes, err := bson.MarshalExtJSON(doc, true, false)
      if err != nil {
      	log.Fatalf("error encoding: %v", err)
      }
      
      m := make(map[string]interface{})
      if err := json.Unmarshal(jsonBytes, &m); err != nil {
      	log.Fatalf("error decoding: %v", err)
      }
      fmt.Printf("json: %v\n", m)

   .. output::
      :language: none
      :visible: false

>       json: map[hello:world]

Reviewer: Question: The description of this code block mentions encoding a BSON document to JSON, but isn't this a Go map? I would have expected JSON to look like:
```json
 { "hello": "world" }
```

Author: good point! changed the explanation and the example and output

> To learn more about encoding and decoding with BSON, see the

Author: changed wording to clarify.

:ref:`golang-bson` guide.

Why Did My Authentication Fail?
-------------------------------

When connecting to a MongoDB deployment that requires password
authentication, you may get an error message if any credentials or the

> authentication configuration are incorrect. For example, the following

Author: It is not only incorrect credentials that can cause an error-- it can also be specifying the authentication method incorrectly. I changed the wording to reflect that an error **will** be surfaced if either of these situations are applicable.

> authentication configuration are incorrect. For example, the following

Author: this makes sense! I didn't understand that you meant to move the conditions to the first clause of the sentence

error message is returned when ``SCRAM-SHA-256`` authentication fails:

.. code-block:: none
   :copyable: false

   connection() error occurred during connection handshake: auth error:
   sasl conversation error: unable to authenticate using mechanism
   "SCRAM-SHA-256": (AuthenticationFailed) Authentication failed.

The error message intentionally omits the exact authentication failure

> reason for security purposes, but common causes for authentication failure include the

Reviewer: Suggestion:
The error message can't have an intention, although the author of the code that generates it can. Avoid personifying inanimate objects when possible. E.g.

This error message does not contain the detailed authentication failure reason for security reasons, but indicates the issue occurred for one of the following reasons:

Author: the error message does not indicate why the issue occurred. changed wording and broke up sentence.

following:

- Incorrect password
- Incorrect username

> - Incorrect specification of authentication method in the

Reviewer: Issue: the Go driver refers to this as authentication mechanism instead of method.

Suggestion:
Slightly less verbose version:

"Incorrect authentication mechanism specified in the AuthSource property"

  ``AuthSource`` property

To avoid this error, ensure that the username and password in your

> connection string or ``SetAuth()`` call are correct. In most cases,

Reviewer: Issue: 

1) The "or SetAuth() call are correct" doesn't seem parallel to "username and password in your connection string", but should be.
2) Avoid recommending putting your credentials in plaintext in the connection string because that is a security risk. If the driver supports providing credentials in environment variables, suggest that instead.

Suggestion:
Start with the general case (credentials are correct) and then describe two specific ways to achieve the general case. E.g. 
To avoid this error, make sure the credentials you provided are correct. You can set your authentication credentials in environment variables or by passing them to the SetAuth() method.

   see the :ref:`golang-issues-and-help` page for next steps and more
   resources.
   
How Can I Fix the "WriteXXX can only write while positioned on a Element or Value but is positioned on a TopLevel" Error?
-------------------------------------------------------------------------------------------------------------------------

The ``bson.Marshal()`` method requires a parameter that can be decoded
into a BSON document, such as the ``bson.D`` type. This error occurs
when the input to ``bson.Marshal()`` is something *other* than a BSON
document.

Situations in which this error can occur include the following:

- You pass a string to ``bson.Marshal()``, causing a ``WriteString`` error.
- You pass a ``null`` to ``bson.Marshal()``, causing a ``WriteNull`` error.
- You pass an integer to ``bson.Marshal()``, causing a ``WriteInt32`` error.

You may encounter this error when you perform a CRUD operation that
internally uses the ``bson.Marshal()`` method or when you call
``bson.Marshal()`` directly to encode data.

The following code produces a ``WriteNull`` error because the driver
cannot encode the ``null`` valued ``sortOrder`` variable to BSON during

> the find-and-update operation:

Reviewer: Suggestion: 
Unless "find-and-update" is a common term for the Go driver (I haven't seen it used in other drivers), it could be more clear to refer to it by the method name instead:

```suggestion
the FindOneAndUpdate() operation:
```

.. code-block:: go

   var sortOrder bson.D
   opts := options.FindOneAndUpdate().SetSort(sortOrder)
   

>    update := bson.D{{"$inc", bson.D{{"counter", 1}}}}

Reviewer: Similar comment to https://github.com/mongodb/docs-golang/pull/226#discussion_r1072938457

Author: changed to `updateDocument`

   result := coll.FindOneAndUpdate(context.TODO(), bson.D{}, update, opts)
   if err := result.Err(); err != nil {
       panic(err)
   }

The following code shows how to correctly initialize the ``sort``
variable as a ``bson.D`` type so that the driver can convert it to BSON:

.. code-block:: go

>    sort := bson.D{}

Reviewer: Similar comment to https://github.com/mongodb/docs-golang/pull/226#discussion_r1072938457

Author: thx forgot to change it here too!

How Do I Convert a BSON Document to JSON?
-----------------------------------------

The driver provides a variety of marshaller methods that can be used to
convert a BSON document to JSON, such as the ``MarshalExtJSON()``
method. To view a readable form of the JSON encoding, you must use
an unmarshaller method or string type-casting to parse the JSON byte

> format.

Reviewer: 👍

The following code converts a BSON document to JSON using the
``MarshalExtJSON()`` method, then parses and prints the JSON byte array
using string type-casting:

.. io-code-block::
   :copyable: true

   .. input::
      :language: go
      :emphasize-lines: 3

      bsonDocument := bson.D{{"hello", "world"}}
      
      jsonBytes, err := bson.MarshalExtJSON(bsonDocument, true, false)
      if err != nil {
          panic(err)
      }
      
      fmt.Println(string(jsonBytes))

   .. output::
      :language: none
      :visible: false

      {"hello":"world"}

> To learn more about conversions between BSON and Go values, see the

Author: I was using the phrasing from the Work with BSON page, but I think that that should be changed there too then. LMK if you think that I should change that in this PR or if it is out of scope.

> To learn more about conversions between BSON and Go values, see the

Reviewer: After looking at the Work with BSON page, there are a couple instances in which "type[s]" would be preferable. I think a ticket would probably be better since that task doesn't seem central to publishing the FAQ.

:ref:`golang-bson` guide.

Why Did the Driver Throw an "Authentication failed" Error?
----------------------------------------------------------

When connecting to a MongoDB deployment that requires password
authentication, you will receive an error message if any credentials or the
authentication configuration are incorrect. For example, the following
error message is returned when ``SCRAM-SHA-256`` authentication fails:

.. code-block:: none
   :copyable: false

   connection() error occurred during connection handshake: auth error:
   sasl conversation error: unable to authenticate using mechanism
   "SCRAM-SHA-256": (AuthenticationFailed) Authentication failed.

This error message does not contain the detailed authentication failure
reason for security reasons. Common causes for authentication failure include the
following:

- Incorrect password
- Incorrect username

> - Incorrect authentication mechanism specified in the ``AuthSource`` property

Reviewer: Question:
Is it only in the AuthSource property or could it be in the connection string?

Author: authSource (lowercase a) is the connection string option, and AuthSource is the property of the Credential struct. I changed the wording to reflect this

To avoid this error, ensure that your credentials and authentication
mechanism are correct. You can store your authentication credentials in
environment variables or you can pass them to the ``SetAuth()`` method.
In most cases, you do not need to set the ``AuthSource`` property unless

> you are using a non-default authentication database.

Reviewer: Suggestion:
This sentence seems to provide information that isn't related to the main topic of why someone received this error message and how to resolve it.

If it seems essential, I think "In most cases" makes it unclear as to what other reasons you would need to set the AuthSource property other than what's listed, and would be helpful to list the other cases.

Author: I think it's best to just omit this sentence, I think this info is given on both authentication fundamentals pages anyways.

into a BSON document, such as the ``bson.D`` type. This error occurs
when you pass something *other* than a BSON document to
``bson.Marshal()``.

The ``WriteNull`` error occurs when you pass a ``null`` to
``bson.Marshal()``. Situations in which a similar error can occur
include the following:

- You pass a string to ``bson.Marshal()``, causing a ``WriteString`` error.
- You pass a boolean to ``bson.Marshal()``, causing a ``WriteBoolean`` error.
- You pass an integer to ``bson.Marshal()``, causing a ``WriteInt32`` error.

You may encounter this error when you perform a CRUD operation that
internally uses the ``bson.Marshal()`` method or when you call
``bson.Marshal()`` directly to encode data.

The following code produces a ``WriteNull`` error because the driver
cannot encode the ``null`` valued ``sortOrder`` variable to BSON during
the ``FindOneAndUpdate()`` operation:

.. code-block:: go

   var sortOrder bson.D
   opts := options.FindOneAndUpdate().SetSort(sortOrder)
   
   updateDocument := bson.D{{"$inc", bson.D{{"counter", 1}}}}
   result := coll.FindOneAndUpdate(context.TODO(), bson.D{}, updateDocument, opts)
   if err := result.Err(); err != nil {
       panic(err)
   }

> The following code shows how to correctly initialize the ``sort``

Reviewer: ```suggestion
The following code shows how to correctly initialize the ``sortOrder``
```

method. To view a readable form of the JSON encoding, you must use
an unmarshaller method or string type-casting to parse the JSON byte
format.

The following code converts a BSON document to JSON using the
``MarshalExtJSON()`` method, then parses and prints the JSON byte array
using string type-casting:

.. io-code-block::
   :copyable: true

   .. input::
      :language: go
      :emphasize-lines: 3

      bsonDocument := bson.D{{"hello", "world"}}
      
      jsonBytes, err := bson.MarshalExtJSON(bsonDocument, true, false)
      if err != nil {
          panic(err)
      }
      
      fmt.Println(string(jsonBytes))

   .. output::
      :language: none
      :visible: false

      {"hello":"world"}

To learn more about conversions between BSON and Go types, see the
:ref:`golang-bson` guide.

Why Did the Driver Throw an "Authentication failed" Error?
----------------------------------------------------------

If you attempt to connect to a MongoDB deployment using incorrect
username and password credentials or specifying the wrong authentication
mechanism, you receive an "Authentication failed" error message. For

> example, the following error message is returned when ``SCRAM-SHA-256``
> authentication fails:

Reviewer: Suggestion: 

Connect the previous sentence about what potential causes are possible with one specific stated cause:

```suggestion
example, the following error message is returned when you
pass invalid credentials when authenticating with ``SCRAM-SHA-256``:
```
